drivers/mdio/mdio_nxp_enet_qos.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_enet_qos_mdio
+
+#include <zephyr/net/mdio.h>
+#include <zephyr/drivers/mdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/ethernet/eth_nxp_enet_qos.h>
+#include <zephyr/sys/util.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mdio_nxp_enet_qos, CONFIG_MDIO_LOG_LEVEL);
+
+struct nxp_enet_qos_mdio_config {
+	const struct device *enet_dev;
+};
+
+struct nxp_enet_qos_mdio_data {
+	struct k_mutex mdio_mutex;
+};
+
+struct mdio_transaction {
+	enum mdio_opcode op;
+	union {
+		uint16_t write_data;
+		uint16_t *read_data;
+	};
+	uint8_t portaddr;
+	uint8_t regaddr;
+	enet_qos_t *base;
+	struct k_mutex *mdio_bus_mutex;
+};
+
+static bool check_busy(enet_qos_t *base)
+{
+	uint32_t val = base->MAC_MDIO_ADDRESS;
+
+	/* Return the busy bit */
+	return ENET_QOS_REG_GET(MAC_MDIO_ADDRESS, GB, val);
+}
+
+static int do_transaction(struct mdio_transaction *mdio)
+{
+	enet_qos_t *base = mdio->base;
+	uint8_t goc_1_code;
+	int ret;
+
+	if (check_busy(base)) {
+		return -EIO;
+	}
+
+	ret = k_mutex_lock(mdio->mdio_bus_mutex, K_FOREVER);
+	if (ret) {
+		goto done;
```

Comment:
You have an error in getting the lock but then route to done where you then do the unlock? Shouldn't this actually do  return some error code? -EAGAIN or something?

Reply:
as a side note, most of zephyr doesn't even check the status of k_mutex_lock(mutex, K_FOREVER). That does seem a bit dangerous and I'm wondering why static analysis doesn't flag this in the other cases.

Reply:
should clean this up then